### ISSUE TITLE
Model description API return 0 memoryUsage at the first a few seconds after the model registered and inference ready

### ISSUE THREAD

**Issue by shihanmax** [pytorch/serve#547]
Your issue may already be reported!
Please search on the [issue tracker](https://github.com/pytorch/serve/issues) before creating one.

## Context

I'm trying to get the memory usage of every worker of a model after I registered it, but the memoryUsage from GET /models/{model_name} turns to be 0 at the first a few seconds (10~30s), even if I already successfully did a batch of inference.  

* torchserve version: 0.1.1
* torch version: 1.4.0
* torchvision version [if any]:
* torchtext version [if any]:
* torchaudio version [if any]:
* java version:java 13-ea
* Operating System and version: Mac os 10.15.5

## Your Environment

* Installed using source? [yes/no]: no
* Are you planning to deploy it using docker container? [yes/no]: no
* Is it a CPU or GPU environment?: CPU
* Using a default/custom handler? [If possible upload/share custom handler/model]: custom 
* What kind of model is it e.g. vision, text, audio?: text
* Are you planning to use local models from model-store or public url being used e.g. from S3 bucket etc.? local
  [If public url then provide link.]:
* Provide config.properties, logs [ts.log] and parameters used for model registration/update APIs:
* Link to your project [if any]:

## Expected Behavior

The model describe API (../models/demo) should be able to give the actual memory usage of the model, at lease when the inference API(../predictions/demo) is ready.

## Current Behavior

The response of requesting model description returns like this:
```
[
  {
    "modelName": "demo",
    "modelVersion": "1.0",
    "modelUrl": "demo.mar",
    "runtime": "python3",
    "minWorkers": 1,
    "maxWorkers": 1,
    "batchSize": 1,
    "maxBatchDelay": 100,
    "loadedAtStartup": false,
    "workers": [
      {
        "id": "9032",
        "startTime": "2020-07-22T20:57:49.934Z",
        "status": "READY",
        "gpu": false,
        "memoryUsage": 0
      }
    ]
  }
]
```
the `memoryUsage` field will back to normal (in my case: 201785344) after a few seconds
## Possible Solution

I'm wondering if there is some refreshing mechanism (say, will update the actual memory usage after a certain amount of time after the model is registered? )

## Steps to Reproduce

1. start torchServe server 
2. register a custom model
3. do inference
4. describe the model through: curl http://localhost:8081/models/demo
5. check the response
6. do step4 and step5 30s later

**Comment by harshbafna** [comment]
@shihanmax : Please share the following :

 - custom handler used [if possible please share the mar file itself]
 - TorchServe logs [logs/ts_log.log]
 - Access logs [logs/access_log.log]

Also, could you please try downgrading the java version to JDK11?

**Comment by shihanmax** [comment]
> @shihanmax : Please share the following :
> 
> * custom handler used [if possible please share the mar file itself]
> * TorchServe logs [logs/ts_log.log]
> * Access logs [logs/access_log.log]
> 
> Also, could you please try downgrading the java version to JDK11?

following is my custom handler:

```
import os
import logging
import torch
import sys
import json

logger = logging.getLogger("handler")
sys.path.append("abspath/to/external/module")

from bert.downstream.bert_for_sequence_classification import (
    BERTForSequenceClassification)
from bert.model.bert.bert import BERT
from bert.dataset.vocab import BERTVocab
from bert.params.params import PretrainParams, GeneralParams, SeqClsParams
from bert.inference.inference import InferenceUtilities


class ModelHandler(object):
    """
    A base Model handler implementation.
    """

    def __init__(self):
        self.error = None
        self._context = None
        self._batch_size = 1
        self.initialized = False

        # customize
        self.model = None
        self.vocab = None
        self.manifest = None
        self.device = None
        self.model_dir = None

    def initialize(self, context):
        """
        Initialize model. This will be called during model loading time
        :param context: Initial context contains model server system properties.
        :return:
        """
        self._context = context
        self._batch_size = context.system_properties["batch_size"]
        self.initialized = True

        self.manifest = context.manifest
        properties = context.system_properties
        self.model_dir = properties.get("model_dir")
        serialized_file = self.manifest['model']['serializedFile']
        model_pt_path = os.path.join(self.model_dir, serialized_file)

        self.device = torch.device("cuda:" + str(properties.get("gpu_id"))
                                   if torch.cuda.is_available() else "cpu")

        self.vocab = BERTVocab(
            raw_data_provider=None,
            vocab_path=PretrainParams.vocab_path,
            force_rebuild=None)

        bert = BERT(vocab_size=len(self.vocab),
                    hidden=PretrainParams.hidden,
                    ffn_hidden=PretrainParams.ffn_hidden,
                    n_attn_head=PretrainParams.attn_heads,
                    n_layers=PretrainParams.layers,
                    max_position_dim=PretrainParams.max_sequence_len,
                    max_seq_len=PretrainParams.max_sequence_len,
                    dropout_rate=PretrainParams.global_dropout_rate)

        model = BERTForSequenceClassification(
            bert_model=bert,
            num_labels=SeqClsParams.num_sequence_labels,
            dropout=GeneralParams.global_dropout_rate
        )

        model.load_state_dict(torch.load(model_pt_path), strict=False)
        model.to(self.device)
        model.eval()
        self.model = model

    def preprocess(self, batch):
        """
        Transform raw input into model input data.
        :param batch: list of raw requests, should match batch size
        :return: list of preprocessed model input data
        """
        # Take the input data and pre-process it make it inference ready
        assert self._batch_size == len(batch), (
            "Invalid input batch size: {}".format(len(batch)))

        logger.info("\n\n\nbatch:{}\n\n\n".format(batch))

        data = batch[0].get("data")
        if data:
            text_list = json.loads(data)
            logger.info("get data:{}".format(text_list))

        else:
            text_list = json.loads(batch[0].get("body"))

        text_list = [i.decode('utf-8') if not isinstance(i, str) else i
                     for i in text_list]

        logger.info("\n\nTEXT LIST: {}\n\n".format(text_list))

        data_loader = InferenceUtilities.prepare_data_for_pattern_clf(
            vocab=self.vocab,
            stem_list=text_list,
            max_seq_len=SeqClsParams.max_seq_len,
            batch_size=self._batch_size
        )

        return data_loader

    def inference(self, model_input):
        """
        Internal inference methods
        :param model_input: transformed model input data, DataLoader
        :return: list of inference output in NDArray:
                    [NDArray(batch_size, num_labels), ..]
        """
        # Do some inference call to engine here and return output
        output_list = []
        logger.info("length of input of inference:{}".format(len(model_input)))
        for data in model_input:
            data = {key: value.to(self.device) for key, value in data.items()}
            output = self.model(
                x=data["token_ids"],
                segment_ids=data["segment_ids"],
                position_ids=data["position_ids"],
                category_ids=data["category_ids"],
                attention_mask=data["attention_mask"]
            )
            output_list.append(output)

        logger.info("length of output of inference:{}".format(len(output_list)))
        return output_list

    def postprocess(self, inference_output):
        """
        Return predict result in batch.
        :param inference_output: list of inference output
        :return: list of predict results
        """
        # Take output from network and post-process to desired format
        max_index = []

        for output in inference_output:
            max_index.append(torch.argmax(output, dim=-1).cpu().numpy().item())

        logger.info("length of output of postprocess:{}".format(len(max_index)))

        return max_index

    def handle(self, data, context):
        model_input = self.preprocess(data)
        model_out = self.inference(model_input)
        return self.postprocess(model_out)


_service = ModelHandler()


def handle(data, context):
    if not _service.initialized:
        _service.initialize(context)

    if data is None:
        return None

    return _service.handle(data, context)
```

[access_log.log](https://github.com/pytorch/serve/files/4963549/access_log.log)
[ts_log.log](https://github.com/pytorch/serve/files/4963550/ts_log.log)

**Comment by shihanmax** [comment]
The problem remains after I downgraded to jdk 11.

```bash
java --version
openjdk 11.0.8 2020-07-14
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.8+10)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.8+10, mixed mode)
```

**Comment by harshbafna** [comment]
@shihanmax: This is observed because the memory information availability is dependent on `TorchServe`'s system metric collector, which runs every 60 seconds. Thus, the model worker's memory information availability could vary anywhere between 0-60 seconds based on when the metric collector gets triggered. 

For TorchServe, this a secondary information. What is the use-case you are trying to implement?

One of the probable use cases in production is auto-scaling based on the system's memory availability. For that, we will be adding [metric endpoints](https://github.com/pytorch/serve/pull/460) in future releases.

**Comment by shihanmax** [comment]
> For TorchServe, this a secondary information. What is the use-case you are trying to implement?
> 
> One of the probable use cases in production is auto-scaling based on the system's memory availability. For that, we will be adding [metric endpoints](https://github.com/pytorch/serve/pull/460) in future releases.

I was planning to do same thing like `auto-scaling`, to be more specific, I want to unregister some models automatically based on their memory usage and recent usage when registering a model with not enough memory. 

Does the `secondary` in your reply means there are more flexible ways to obtain the memory informations?

**Comment by harshbafna** [comment]
> > For TorchServe, this a secondary information. What is the use-case you are trying to implement?
> > One of the probable use cases in production is auto-scaling based on the system's memory availability. For that, we will be adding [metric endpoints](https://github.com/pytorch/serve/pull/460) in future releases.
> 
> I was planning to do same thing like `auto-scaling`, to be more specific, I want to unregister some models automatically based on their memory usage and recent usage when registering a model with not enough memory.

Generally, I would not unregister/scale down a model based on their memory usage but based on the inference load on that model. Correct me if I am wrong, but in the case of auto-scaling, you would be adding more resources to your already running server rather then unregistering/scaling down a model? 

Also, do you envision a use-case where you will need to unregister a model within 60 seconds of its registration based on its memory consumption? As explained in the previous comment, the memory consumed by the model worker will be available within 60 seconds.

> 
> Does the `secondary` in your reply means there are more flexible ways to obtain the memory informations?

The idea was that the auto-scaling will be based on the server's resources not based on a model's memory consumption.

**Comment by shihanmax** [comment]
@harshbafna  Thanks for the explanation, My application needs to consider the situation of many models and limited resources (all models cannot be loaded at the same time).

Here is another question: I encountered a situation here. When I tested a model, `nvidia-smi` indicated the memory occupation of the worker is 1013M, and the MemoryUsage in the Describe model API is 2345M. At this time, the memory usage of the corresponding pid is 2246M, the question is,  in the Describe Model API, does the `memoryUsage` of each worker refers to the usage of memory or the memory of GPU?

My requirement is to obtain the memory usage and GPU memory usage of each worker. Is there any appropriate way to obtain these two indicators? (For example, can I get the pid of given worker, and then manually obtain the above indicators through it?)

**Comment by harshbafna** [comment]
> Here is another question: I encountered a situation here. When I tested a model, `nvidia-smi` indicated the memory occupation of the worker is 1013M, and the MemoryUsage in the Describe model API is 2345M. At this time, the memory usage of the corresponding pid is 2246M, the question is, in the Describe Model API, does the `memoryUsage` of each worker refers to the usage of memory or the memory of GPU?

> My requirement is to obtain the memory usage and GPU memory usage of each worker. Is there any appropriate way to obtain these two indicators? (For example, can I get the pid of given worker, and then manually obtain the above indicators through it?)

`TorchServe` uses [psutil](https://pypi.org/project/psutil/) to capture the CPU and memory (RAM) usage and does not capture the GPU memory usage of a model.

Could you please raise an enhancement ticket to include the GPU memory information and worker pid in `describe model` api response?

If this is a blocker for you, then you can try out following hack in your model's handler:

 - Add a hack in handler's request function to either return the required information (GPU memory) or run the inference based on the input received in the inference request

 - You can use [nvgpu](https://pypi.org/project/nvgpu/) to collect the GPU memory information

 - The `nvgpu` package is not installed by default along with TorchServe.  You can use the new feature of TorchServe which allows users to supply a model-specific requirements.txt file containing a list of 3rd-party python dependency packages. Refer to [documentation](https://github.com/pytorch/serve/blob/master/docs/configuration.md#allow-model-specific-custom-python-packages) for more details. Note that this feature is only available in the latest master as of now and will be available in the next release.

**Comment by harshbafna** [comment]
Closing this ticket as the original question has been answered. The enhancement to describe API will be tracked through #597.

**Comment by shihanmax** [comment]
Thank you for your patience and look forward to the following optimization!
